### PR TITLE
[READY] Warm hugs for everyone

### PR DESCRIPTION
**Pull request by nightred**
## About The Pull Request

If you ever feel down and out, a nice warm hug will brighten your day.
When you hug someone you warm the lowest body temp is warmed up. This accounts for insulation, max temperature change rates, and change steps.

Further changes
* proc natural_bodytemperature_stabilization is now self contained,
* proc adjust_bodytemperature has been expanded for carbon mobs to allow the use of insulation and change steps
* most of handle_environment has been moved to natural_bodytemperature_stabilization and adjust_bodytemperature cleaning it up.
* When hugging both parties temperature is adjusted one up one down, you can not freeze someone without also freezing yourself. 

## Why It's Good For The Game

Some time you need a warm hug to get you through the day.
Save someone freezing to death by hugging them.
Ethereal's are nice and warm think of them when you look for that next hug.
Code cleanup round two on handle environment, separating parts out of the old large function

## Changelog
:cl:
add: Hugs are warm now
refactor: natural_bodytemperature_stabilization is now self contained.
refactor: adjust_bodytemperature to allow the use of insulation and change steps
/:cl:

### CODE REVIEW

**Comment by HomocidlePancakes** [comment]
Hugs reducing stamcrit recovery time when?

**Comment by GuyonBroadway** [comment]
What if someone is running a feaver from a virus, could you in theory heat people beyond standard?

**Comment by HomocidlePancakes** [comment]
Break the laws of thermodynamics with hugs.

**Comment by XDTM** [comment]
Xenobiologist can now spacewalk while being hug-warmed by their gang of golems

**Comment by zxaber** [comment]
Does this work for borgs? I look forward to a borg sitting for a few minutes in a fire and then hugging people to burn them.

**Comment by cacogen** [comment]
Should be able to throw eggs at hot borgs to make fried eggs.

**Comment by nightred** [comment]
Borg's are not included.

**Comment by nightred** [comment]
> What if someone is running a feaver from a virus, could you in theory heat people beyond standard?

You can warm people beyond normal body temp, but if you are over the heat damage limit (or they are) you do not want to hug them. This will prevent burns.

**Comment by XDTM** [comment]
I wouldn't restrict hugging by temperature, maybe simply add a message when it happens so people know what made them overheat in case they hug a recently-burning person.

**Review by Jalleo — COMMENTED** [review]
I hope this doesn't work through rigsuits , firesuits , (bio/rad/bomb)suits , spacesuits and more

**Comment by nightred** [comment]
> I wouldn't restrict hugging by temperature, maybe simply add a message when it happens so people know what made them overheat in case they hug a recently-burning person.

I removed the block on the hug and changed it to a output message.
I also added a new mood when you are warmed up by a hug.

> I hope this doesn't work through rigsuits , firesuits , (bio/rad/bomb)suits , spacesuits and more

Because this checks for insulation you are not warmed up when wearing full insulation, and the more insulation you have the worse it works.

File: code/datums/mood_events/generic_positive_events.dm [hunk 1]
```diff
@@ -19,6 +19,14 @@
 /datum/mood_event/besthug/add_effects(mob/friend)
 	description = "<span class='nicegreen'>[friend.name] is great to be around, [friend.p_they()] makes me feel so happy!</span>\n"
 
+/datum/mood_event/warmhug
+	description = "<span class='nicegreen'>Warm cozy hugs are the best!</span>\n"
+	mood_change = 1
+	timeout = 2 MINUTES
+
+/datum/mood_event/warmhug/add_effects()
+	description = "<span class='nicegreen'>Someone gave me a warm cozy hug!</span>\n"
```
Comment (Mickyan) [inline code/datums/mood_events/generic_positive_events.dm]: what's this for?

Comment (nightred) [reply code/datums/mood_events/generic_positive_events.dm]: When you click the moodlet icon it reports this back to you. Am I wrong?

Comment (Mickyan) [reply code/datums/mood_events/generic_positive_events.dm]: that's what the description in **/datum/mood_event/warmhug** does

You only need this proc if you, say, want the message to include the name of the person that hugged you, like in **/datum/mood_event/besthug**

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 2]
```diff
@@ -267,7 +267,28 @@
...
+
+		else // you are warm share the heat of life
+			var/temp_diff = M.bodytemperature - bodytemperature
+			adjust_bodytemperature(min((1 - get_insulation_protection(M.bodytemperature)) * temp_diff / BODYTEMP_HEAT_DIVISOR, \
+			BODYTEMP_HEATING_MAX))
+			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/warmhug, M)
+
+		// Let people know if they hugged someone really warm
+		if(M.bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>[M] felt like they are over heating when they hugged you.</span>")
```
Comment (ATH1909) [inline code/modules/mob/living/carbon/carbon_defense.dm]: tenses

**Comment by ATH1909** [comment]
This should cool down/heat up the person performing the hug as well. Something something laws of thermodynamics.

Can you cool yourself down to near 0K as a plasmaman, then hug lizardpeople to kill them with cold?

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 3]
```diff
@@ -267,7 +267,28 @@
...
+
+		if(bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
```
Comment (ATH1909) [inline code/modules/mob/living/carbon/carbon_defense.dm]: add messages for hugging/being hugged by people who are freezing too

**Comment by ATH1909** [comment]
> Does this work for borgs? I look forward to a borg sitting for a few minutes in a fire and then hugging people to burn them.

I doubt it; their hugging modules don't even help people recover from being stunned or knocked unconscious faster.

File: code/modules/mob/living/carbon/carbon_defense.dm (hunk 3 above)
Comment (nightred) [reply code/modules/mob/living/carbon/carbon_defense.dm]: done

File: code/modules/mob/living/carbon/carbon_defense.dm (hunk 2 above)
Comment (nightred) [reply code/modules/mob/living/carbon/carbon_defense.dm]: think i fixed it up.

**Comment by nightred** [comment]
> This should cool down/heat up the person performing the hug as well. Something something laws of thermodynamics.

You are now cooled when you hug someone. Insulation and difference range, and max heating cooling steps are still accounted for.

> Can you cool yourself down to near 0K as a plasmaman, then hug lizardpeople to kill them with cold damage?

Sort of, the warm person will cool down, the cold person will warm up. At 0 kelvin a human at 310 kelvin will heat you up about 20 degrees k per hug, and you will cool down about that much also.

> Does this work for borgs? I look forward to a borg sitting for a few minutes in a fire and then hugging people to burn them.

This only work on mob/living/carbon/ sub types, borgs are part of mob/living/silicon and can not directly interact with carbons the same way.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 4]
```diff
@@ -267,7 +267,40 @@
...
+			adjust_bodytemperature(min((1 - get_insulation_protection(M.bodytemperature)) * temp_diff / BODYTEMP_HEAT_DIVISOR, \
+			BODYTEMP_HEATING_MAX)) // warm up the reciver
+			M.adjust_bodytemperature(min((1 - M.get_insulation_protection(bodytemperature)) * temp_diff / BODYTEMP_COLD_DIVISOR, \
+			BODYTEMP_COOLING_MAX)) // cool down the giver
+
+			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/warmhug, M)
+
+		// Let people know if they hugged someone really warm or really cold
+		if(M.bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>It feels like [M] is over heating when they hug you.</span>")
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/carbon_defense.dm]: overheating as they hug you

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 5]
```diff
@@ -267,7 +267,40 @@
...
+		else if(M.bodytemperature < BODYTEMP_COLD_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>It feels like [M] is freezing when they hug you.</span>")
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/carbon_defense.dm]: as

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 6]
```diff
@@ -267,7 +267,40 @@
...
+
+		if(bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(M, "<span class='warning'>It feels like [src] is over heating when you hug them.</span>")
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/carbon_defense.dm]: .

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 7]
```diff
@@ -267,7 +267,40 @@
...
+		else if(bodytemperature < BODYTEMP_COLD_DAMAGE_LIMIT)
+			to_chat(M, "<span class='warning'>It feels like [src] is freezing when you hug them.</span>")
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/carbon_defense.dm]: .

File: code/modules/mob/living/carbon/life.dm [hunk 8]
```diff
@@ -631,6 +631,20 @@ GLOBAL_LIST_INIT(ballmer_windows_me_msg, list("Yo man, what if, we like, uh, put
...
+/**
+ * Get the insulation that is appropriate to the temperature you're being exposed to.
+ * All clothing, natural insulation, and traits are combined returning a single value.
+ *
+ * required temperature The Temperature that you're being exposed to
+ *
+ * return the percentage of protection as a value from 0 - 1
+**/
+/mob/living/carbon/proc/get_insulation_protection(temperature)
+	if (temperature > bodytemperature) // exposed to heat, get heat protection
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/life.dm]: condense this as so

return (temperature > bodytemperature) ? get_heat_protection(temperature) : get_cold_protection(temperature)

File: code/modules/mob/living/carbon/carbon_defense.dm (hunk 4 above)
Comment (nightred) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Corrected thanks

File: code/modules/mob/living/carbon/carbon_defense.dm (hunk 5 above)
Comment (nightred) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Correcting also.

File: code/modules/mob/living/carbon/life.dm (hunk 8 above)
Comment (nightred) [reply code/modules/mob/living/carbon/life.dm]: Changed also

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 9]
```diff
@@ -267,7 +267,40 @@
 	else
 		M.visible_message("<span class='notice'>[M] hugs [src] to make [p_them()] feel better!</span>", \
 					"<span class='notice'>You hug [src] to make [p_them()] feel better!</span>")
-		SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/hug)
+
+		// Warm them up with hugs
+		if(bodytemperature > M.bodytemperature) // they are warmer leech from them
```
Comment (Cyberboss) [inline code/modules/mob/living/carbon/carbon_defense.dm]: This feels like copypasta from the temperature PR. If I'm right, can we make them share the code somehow please?

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/carbon_defense.dm (hunk 9 above)
Comment (nightred) [reply code/modules/mob/living/carbon/carbon_defense.dm]: This is not a direct copy, but sharing body heat might be good to get as a proc somewhere in the mob framework. I have an idea let me see.

Comment (nightred) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Ok Cyber I did some major changes.
natural_bodytemperature_stabilization is now self contained.
adjust_bodytemperature is expanded for carbon mobs to allow insulation and change steps
handle_environment has been reduced in size as most of it is now done in the other two steps.

**Comment by ExcessiveUseOfCobblestone** [comment]
they should share heat instead of getting "free" heat.

**Comment by nightred** [comment]
> they should share heat instead of getting "free" heat.

They do share heat now actually, in the code that is. i guess i did not document that, fixing.

**Comment by nightred** [comment]
Issue still with hard suits

**Comment by GuyonBroadway** [comment]
Idea for a followup pr.

Hugs spread radiation for extra spicy hugging.

**Comment by nightred** [comment]
> Idea for a followup pr.
> Hugs spread radiation for extra spicy hugging.

Funny thought but just being near someone with rads will give you rads so no need to do anything extra. Do not hug glowers.

File: code/modules/mob/living/carbon/life.dm [hunk 10]
```diff
@@ -679,15 +679,22 @@ GLOBAL_LIST_INIT(ballmer_windows_me_msg, list("Yo man, what if, we like, uh, put
...
-	// use the bodytemp divisors to get the change step, with max step size
+	// Extra calculation for hardsuits to bleed off heat
+	if(hardsuit_fix)
+		amount += hardsuit_fix
+
+	// Use the bodytemp divisors to get the change step, with max step size
 	if(use_steps)
-		amount = (amount > 0) ? min(amount / BODYTEMP_HEAT_DIVISOR, BODYTEMP_HEATING_MAX) : max(amount / BODYTEMP_COLD_DIVISOR, BODYTEMP_COOLING_MAX)
+		amount = (amount > 0) ? min(amount / BODYTEMP_HEAT_DIVISOR, BODYTEMP_HEATING_MAX) : \
+			max(amount / BODYTEMP_COLD_DIVISOR, BODYTEMP_COOLING_MAX)
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/life.dm]: 1 line please

File: code/modules/mob/living/carbon/human/species.dm [hunk 11]
```diff
@@ -1633,7 +1633,8 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 		natural_bodytemperature_stabilization(environment, H)
 
 	if(!H.on_fire || areatemp > H.bodytemperature) // If we are not on fire or the area is hotter
-		H.adjust_bodytemperature((areatemp - H.bodytemperature), use_insulation=TRUE, use_steps=TRUE)
+		H.adjust_bodytemperature((areatemp - H.bodytemperature), use_insulation=TRUE, use_steps=TRUE, \
+			hardsuit_fix=bodytemp_normal - H.bodytemperature)
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/human/species.dm]: same here

Comment (nightred) [reply code/modules/mob/living/carbon/human/species.dm]: fixed this up to be wide style

File: code/modules/mob/living/carbon/life.dm (hunk 10 above)
Comment (nightred) [reply code/modules/mob/living/carbon/life.dm]: fixed this up to be wide style

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 12]
```diff
@@ -267,7 +267,26 @@
...
 					"<span class='notice'>You hug [src] to make [p_them()] feel better!</span>")
-		SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/hug)
+
+		// Warm them up with hugs
+		share_bodytemperature(M)
+		if(bodytemperature > M.bodytemperature)
+			SEND_SIGNAL(M, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/warmhug, src) // Hugger got a warm hug
+			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/hug) // Reciver always gets a mood for being hugged
+		else
+			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/warmhug, M) // You got a warm hug
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/carbon_defense.dm]: why don't you just unindent

Comment (nightred) [reply code/modules/mob/living/carbon/carbon_defense.dm]: So that I do not overwrite a mood set one before, and only the cold hugger gets the warm mood.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/carbon/carbon_defense.dm]: I literally saw this and don't know why I still sent the comment cringe

File: code/modules/mob/living/carbon/human/species.dm [hunk 13]
```diff
@@ -1782,31 +1748,61 @@ GLOBAL_LIST_EMPTY(roundstart_races)
...
-/// Returns the amount of degrees kelvin to change the body temperature
-/datum/species/proc/natural_bodytemperature_stabilization(mob/living/carbon/human/H)
+/**
+ * Used to stabilize the body temperature back to normal on living mobs
+ *
+ * vars:
+ * * environment The environment gas mix
+ * * H The mob we will stabilize
+ */
+/datum/species/proc/natural_bodytemperature_stabilization(datum/gas_mixture/environment, mob/living/carbon/human/H)
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/human/species.dm]: is this a duplicate of the carbon still? they look awfully similar

Comment (nightred) [reply code/modules/mob/living/carbon/human/species.dm]: This version uses the species body temperature customizations.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/carbon/human/species.dm]: rad

**Pull request merged**